packages/smooth_app/lib/pages/offline_data_page.dart

```diff
@@ -0,0 +1,274 @@
+                                  daoProduct, daoProductList),
+                              title: 'Refreshing \n This may take a while',
+                              dismissible: true,
+                            ) ??
+                            'Refresh Cancelled';
+                        if (!mounted) {
+                          return;
+                        }
+                        ScaffoldMessenger.of(context).showSnackBar(
+                          SnackBar(
+                            content: Text(status),
+                          ),
+                        );
+                      },
+                      title: const Text(
+                        'Update Offline Data',
+                      ),
+                      subtitle: const Text(
+                        'Update the local databse cache with the latest data from the server',
+                      ),
+                    ),
+                    UserPreferencesListTile(
+                      trailing: const Icon(Icons.delete),
+                      onTap: () async {
+                        //show a dialog to confirm the deletion
+                        final bool confirmed = await showDialog<bool>(
+                              context: context,
+                              builder: (BuildContext context) {
+                                return SmoothAlertDialog(
+                                  title: 'Confirm Deletion?',
+                                  body: const Text('''
+                                      It will delete all the cached products as well as the scan history\n
+                                      This process cannot be reversed, Do you wish to continue?'''),
+                                  positiveAction: SmoothActionButton(
+                                    onPressed: () async {
+                                      Navigator.pop(context, true);
+                                    },
+                                    text: 'Okay',
+                                  ),
+                                  negativeAction: SmoothActionButton(
+                                    onPressed: () async {
+                                      Navigator.pop(context, false);
+                                    },
+                                    text: 'Cancel',
+                                  ),
+                                );
+                              },
+                            ) ??
+                            false;
+                        if (confirmed) {
+                          final double size = await localDatabase.getSize();
+                          final List<String> typesToDelete =
+                              await daoProductList.getKeys();
+                          // just keep the product list keys that should be deleted
+                          // ignore: list_remove_unrelated_type
+                          typesToDelete.remove(!typesToDelete
```

Comment:
That's very ugly. You should use `DaoProductList.getProductListType` instead, or create a method in `DaoProductList` that returns all the product list of the specified types.